OCD-4490: Add code sets, RWT to upload

kekey1 opened the pull request.

kekey1 committed:
feat!: Parse code sets from upload file

[#OCD-4490]

kekey1 committed:
Merge branch 'staging' into OCD-4490

kekey1 committed:
Merge branch 'staging' into OCD-4490

kekey1 committed:
feat!: Support uploading and confirming Code Sets on a listing

I decided to mainly give warnings and remove the code sets for the user
when an invalid one was detected. I made this decision because in most
cases, we don't give the user a way to remove them in the UI. If they
have entered a code set that doesn't exist or that hasn't become
available yet they have no way to remove them. So I'm giving a warning
and removing it for them.

Added validation for duplicate code sets.

Also added validation for a code set that is required (today is after
it's required date) but it's not found on the listing. This is the one
case where we give the user an error and force them to do something
because they can and should.

Commit includes unit tests.

[#OCD-4490]

kekey1 committed:
feat: Allow a few other code set date formats in upload

12-2025, Dec 2025, December 2025

[#OCD-4490]

kekey1 committed:
feat!: Support RWT url and date fields in listing upload

[#OCD-4490]

kekey1 committed:
feat: Correct RWT parsing; save RWT data on confirm

[#OCD-4490]

kekey1 committed:
feat: Add additional accepted date formats for RWT, Code Set parsing

[#OCD-4490]

tmy1313 commented on chpl/chpl-service/src/main/java/gov/healthit/chpl/upload/listing/normalizer/CodeSetNormalizer.java:
```diff
@@ -0,0 +1,117 @@
+        Optional<CodeSet> codeSetOpt = codeSetsForCriterion.stream()
+            .filter(codeSet -> matchesCodeSetName(codeSet, codeSetText))
+            .findAny();
+        return codeSetOpt.isPresent() ? codeSetOpt.get() : null;
```
This can be simplified... 
```
 return codeSetsForCriterion.stream()
            .filter(codeSet -> matchesCodeSetName(codeSet, codeSetText))
            .findAny()
            .orElse(null);
```

kekey1 replied:
Updated

tmy1313 commented on chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/listing/reviewer/RealWorldTestingReviewer.java:
```diff
@@ -55,13 +48,7 @@ private boolean isListingCurrentlyRwtEligible(CertifiedProductSearchDetails list
-            LOGGER.error("Could not determine the edition of listing {}", listing.getId(), e);
-            return false;
-        }
+        return listing.getEdition() == null || listing.getEdition().getName().equals(EDITION_2015);
```
Since we touched this method - should it also be renamed?  Also, just want to confirm that we want to "trust" the `edition` value passed into the reviewer?  Might not matter but just wanted to bring it up.

kekey1 replied:
Updated

kekey1 committed:
refactor: Simplify logic returning null if optional is empty

[#OCD-4490]

andlar commented on chpl/chpl-service/src/main/java/gov/healthit/chpl/upload/listing/normalizer/ListingDetailsNormalizer.java:
```diff
@@ -47,6 +48,7 @@ public ListingDetailsNormalizer(CertificationEditionNormalizer editionNormalizer
         this.editionNormalizer = editionNormalizer;
         this.acbNormalizer = acbNormalizer;
         this.atlNormalizer = atlNormalizer;
+        this.rwtNormalizer = new RwtNormalizer();
```
Why is this the only one that's marked with `new`?

kekey1 replied:
It doesn't have dependencies on any of our other spring components (DAOs, Services, etc), so it's not a spring component (with Autowired constructor and Component annotation). Spring won't create it so I have to.

kekey1 committed:
refactor: Remove edition check for RWT validation

The validators only run against Active Certificates which are all 2015
and beyond edition listings. Any non-active certificate (non-2015
edition) doesn't get validated so this was essentially dead code.

[#OCD-4490]

kekey1 committed:
feat!: Support more criteria headings during upload

Allow both CRITERIA_170_315_D_4__C and CRITERIA_170_315_D_4_C. Users
were getting errors when missing the extra underscore.

[#OCD-4490]

kekey1 committed:
Merge branch 'staging' into OCD-4490

kekey1 committed:
Merge branch 'staging' into OCD-4490

kekey1 committed:
fix: Handle empty code sets list

[#OCD-4490]

kekey1 committed:
fix: Handle code set with null id in reviewer

[#OCD-4490]

kekey1 committed:
Merge branch 'staging' into OCD-4490

kekey1 committed:
fix: Update regex for criteria number parsing

[#OCD-4490]

kekey1 committed:
feat: Add yy-MMM as available code set format

[#OCD-4490]
